File: backend/app/repositories/testcase_repo.py
from sqlalchemy.orm import Session
from typing import List, Optional
from app.models.testcase import TestCase

class TestCaseRepository:
    def __init__(self, db: Session):
        self.db = db

    def get_by_id(self, tc_id: int) -> Optional[TestCase]:
        return self.db.query(TestCase).filter(TestCase.id == tc_id).first()

    def get_by_question(self, question_id: int) -> List[TestCase]:
        return self.db.query(TestCase).filter(TestCase.question_id == question_id).all()

    def create(self, testcase: TestCase) -> TestCase:
        self.db.add(testcase)
        self.db.commit()
        self.db.refresh(testcase)
        return testcase

    def delete(self, testcase: TestCase) -> None:
        self.db.delete(testcase)
        self.db.commit()
        
    def save(self) -> None:
        self.db.commit()

File: backend/app/repositories/testcase_repo.js
const db = require('../database/session');

class TestCaseRepository {
  static async getById(id) {
    const col = await db.collection('testcases');
    return await col.findOne({ _id: id }) || null;
  }

  static async getByQuestion(questionId) {
    const col = await db.collection('testcases');
    return await col.find({ question_id: questionId }).toArray();
  }

  static async create(tc) {
    const col = await db.collection('testcases');
    const id = await db.getNextSequenceValue('testcase_id');
    const doc = {
      _id: id,
      id: id,
      question_id: tc.question_id,
      input: tc.input,
      expected_output: tc.expected_output,
      is_hidden: tc.is_hidden ?? false,
      created_at: new Date()
    };
    await col.insertOne(doc);
    return doc;
  }

  static async delete(id) {
    const col = await db.collection('testcases');
    const res = await col.deleteOne({ _id: id });
    return res.deletedCount > 0;
  }
}

module.exports = TestCaseRepository;
